\appendix

\section{Appendix}

\subsection{Notation}\label{app:notation} For a graph $G$
we write $V(G)$ and $E(G)$ for its vertex and edge set, respectively. For a
vertex $v \in V(G)$, we write $N(v)$ to denote its neighborhood, \ie the set
containing all vertices that are adjacent to it in $G$. We further
write $G-v$ for the graph obtained from $G$ by removing $v$ and all edges
incident to it.

For a directed graph (digraph) $\dir G$, we use $N^-(v)$ to denote
the \emph{in-neighborhood} of $v \in V(\dir G)$, that is, the set 
of all vertices $w$ for which an arc $wv$ exists in $\dir G$. The
size of that in-neighbourhood is called the \emph{in-degree} of $v$.
We write $\maxindeg(\dir G)$ for the maximum in-degree over all 
vertices of $\dir G$.

For ease of notation, we usually write ``$v \in G$'' instead of 
``$v \in V(G)$'' and ``$uv \in G$'' instead of ``$\{u,v\} \in E(G)$''
(same for digraphs).

\subsection{Dtf-augmentation}\label{app:dtf} For the sake
of completeness, we briefly describe how dtf-augmentations
$\dir G_1, \dir G_2, \ldots, \dir G_r$ are computed for an input
graph $G$. 

First, the graph $\dir G_1$ is a \emph{minimum in-degree
orientation} of $G$ obtained as follows: we find a vertex~$u$ of
minimal degree in $G$, orient the incident edges towards $u$ and
the repeat the procedure in $G - u$ until no vertices remain.

Given $\dir G_i$, $i \geq 1$ we then compute $\dir G_{i+1}$ by
applying an \emph{augmentation step}:

\begin{enumerate}
    \item $\dir G_{i+1}$ contains all arcs of $\dir G_i$,
    \item if there is a pair of arcs~$uv \in \dir G_j$ 
          and $vw \in \dir G_k$ with $j + k = i + 1$, then
          we add $uw$ to $\dir G_{i+1}$,
    \item if there is a pair of arcs~$uv \in \dir G_j$ 
          and $wv \in \dir G_k$ with $j + k = i + 1$, then
          either the arc $uw$ or the arc $wu$ must be
          in $\dir G_{i+1}$.
\end{enumerate}
The ambiguity in the last step is resolved as follows: 
we collect all edges $uw$ for which the last case applies
in a graph $H$ and then compute $\dir H = \texttt{orient}(H)$,
after which we add the arcs in $\dir H$ to $\dir G_{i+1}$.

For ease of notation, we define $\omega(uv)$ to be the
smallest index $i \leq r$ such that $uv \in \dir G_i$
(and $\infty$ if no such index exists).
We make use of two crucial properties of dtf augmentations 
in the following. First, vertices that have distance at most
$r$ in $G$ have distance at most two in $\dir G_r$:

\begin{lemma}[\cf Lemma 26 in~\cite{felixThesis}]
  For every pair $u,v \in G$ with $\dist(u,v) = d$ and
  for every integer~$r \geq d$ one of the following holds:
  \begin{enumerate}
    \item $uv \in \dir G_r$ and $\omega(uv) = d$,
    \item $vu \in \dir G_r$ and $\omega(vu) = d$,
    \item or there exists $x$ such that $xu, xv \in \dir G_r$ and
          $\omega(xu) + \omega(xv) = d$.
  \end{enumerate}
\end{lemma}

\noindent
Second, if we compute dtf-augmentations of graphs from a bounded
expansion class (to which graph of bounded maximum degree like
cDBGs belong), then the maximum in-degree
of its $r$th augmentation can be bounded by a function independent
of the \emph{size} of the graph:

\begin{theorem}[\cf Theorem 16 in~\cite{felixThesis}]\label{thm:dtf}
  For every graph class $\mathcal G$ with bounded expansion there exists
  a function~$f$ such that for every member $G \in \mathcal G$
  the $r$th dtf-augmentation $\dir G_r$ computed as above
  satisfies
  $\maxindeg(\dir G_r) \leq f(r)$.
\end{theorem}

\subsection{Approximation guarantee}\label{app:threshold}
% Note: labels don't work right now because we're not numbering subsections (style file issue)

Let us introduce some notation for the analysis of
Algorithm~\ref{alg:r-domset}. We first partition the vertices of~$D$ according
to whether they were added in line~\ref{algstep:D1} (denoted by~$D_1$) or in
line~\ref{algstep:D2} (denoted by~$D_2$). Let~$v_1,\ldots,v_n$ be the vertex-
order in which they are iterated over in the loop starting at
line~\ref{algstep:order}. We will use the notation~$D_1^i$, $D_2^i$, $d^i$,
and~$c^i$ to represent the states of the respective sets and data structures
during the $i$th iteration of said loop. Let~$\tau :=
\texttt{domThreshold}(r)$ be the chosen threshold (we discuss a good value
for $\tau$ on cDBGs below).

\begin{lemma}
  After the for-loop at line~\ref{algstep:pull-loop} has finished,
  \[
    d^i[v_i] = \begin{cases*}
      \dist_G(v,D^i) & if $\dist_G(v,D^i) \leq r$, and \\
      \infty & otherwise.
    \end{cases*}
  \]
\end{lemma}
\begin{proof}
  The statement trivially holds while~$D^i = \emptyset$, so assume otherwise.
  Let~$u_h \in D^i$ be the vertex closest to~$v_i$ and let~$h < i$
  be the iteration in which~$u_h$ was added to~$D$ (either in
  line~\ref{algstep:D1} or line~\ref{algstep:D2} of that iteration).

  If $d := \dist_G(v_i,u_h) > r$,
  then~$d^i[v_i]$ has not been changed yet and is still set to~$\infty$.
  Otherwise, consider the three possible scenarios promised by the
  distance-property of dtf-augmentations:

  \begin{case}[$v_iu_h \in \dir G_d$]
    Then~$\omega(v_iu_h) = d$ and in iteration~$h$
    the value of~$d^h[v_i]$ is set to the correct value~$d$
    at line~\ref{algstep:pull}.
    By assumption this distance remains minimal until iteration~$i$
    and hence~$d^i[v_i] = d^h[v_i] = d$.
  \end{case}

  \begin{case}[$u_hv_i \in \dir G_d$]
    Then~$\omega(v_iu_h) = d$ and in iteration~$i$
    the value of~$d^i[v_i]$ is set to the correct value~$d$
    at line~\ref{algstep:pull}.
  \end{case}

  % The cases env is a bit hacky. Parenthesis don't work well,
  % unless hidden in a macro.
  \def\ww#1{\omega(#1)}
  \begin{case}[$xu_h,xv_i \in \dir G_d$  with~$\ww{xu_h} + \ww{xv_i} = d$]\mbox{}\\
    During iteration~$h$ the value of~$d^h[x]$ is set to~$\ww{xu_h}$
    at line~\ref{algstep:pull} and subsequently retrieved
    in iteration~$i$ when $d^i[v_i]$ is set to
    \[
      d^i[x] + \ww{xu_h} = \ww{xu_h} + \ww{xv_i} = d.
    \]
    \vspace*{-1.5em}
  \end{case}
  \noindent
  We conclude that after the execution of the loop at line~\ref{algstep:pull}.
  $d^i[v_i]$ is set to~$\infty$ if~$v_i$ is not dominated by~$D_i$ and is
  otherwise set to~$\dist_G(v_i,D^i)$, as claimed.
\end{proof}

\noindent
As an immediate consequence, we see the conditional statement at the end of
the loop at line~\ref{algstep:pull} accurately determines whether $v_i$ is
dominated by~$D_i$ or not. Accordingly, line~\ref{algstep:D2}
of the loop is only executed if~$v_i$ is \emph{not} dominated by~$D^i$.
Another consequence is that all vertices in~$D_1$ have large distance to
each other:

\begin{corollary}
  The set~$D_1$ is $(r+1)$-scattered in~$G$.
\end{corollary}

\noindent
We need one more important property of the algorithm in order to derive
the approximation factor.

\begin{lemma}\label{lemma:D1-intersection}
  For every~$w \in G$ it holds that~$|D_1 \cap N^-_r(w)| \leq \tau+1$.
\end{lemma}
\begin{proof}
  Assume towards a contradiction that~$\tau+2$ such vertices
  $v_{i_1},\ldots,v_{i_\tau+2}$, $i_1 < i_2 < \ldots < i_{\tau+2}$ exist in~$D_1
  \cap N^-_r(w)$. Since every such vertex~$v_i$, $i \in \{i_1,\ldots
  i_{\tau+2}\}$, was added to~$D$ in part~\MARK{2}, part~\MARK{3} of the
  algorithm was executed during iteration~$i$ as well. Thus~$c[w]$ was
  increased in each iteration~$i$ and during iteration~$i_{\tau+1}$ we have that
  $c[w] \geq \tau + 1$ after the increment of~$c[w]$. Therefore part~\MARK{4}
  must have been executed for~$w$, including~$w$ into~$D$. Hence~$w \in D^s$
  for~$s > i_{\tau+1}$ and in particular $w \in D^{i_{\tau+2}}$. But then
  $v_{i_{\tau+2}}$ was dominated by~$w$ at the beginning of iteration~$i_{\tau+2}$
  since we assumed that~$\omega(rv_{i_{\tau+2}}) \leq r$, thus~$v_{i_{\tau+2}}$
  would not have been included in~$D$ at step~\MARK{2}. This contradicts our
  assumption of $v_{i_{\tau+2}} \in D_1$ so the claim must hold.
\end{proof}

\begin{lemma}
  There exists a subset~$A \subseteq D_1$ such that~$A$ is
  $(2r+1)$-scattered in~$G$ and
  \[
  	|A| \geq \frac{|D|}{2(\tau+2)\maxindeg(\dir G_{2r}))\maxindeg(\dir G_r)}.
  \]
\end{lemma}
\begin{proof}
  We construct an auxiliary graph~$H$ with vertices~$D_1$ by
  adding arcs~$v_iv_j$ for $v_i,v_j \in D_1$ with $i < j$
  whenever~$\dist_G(v_i,v_j) \leq 2r$.
  Let~$\dir G_{2r}$ be a $2r$th dtf-augmentation of~$G$ and
  let us create a digraph~$\dir H$ by orienting
  every edge~$uv \in H$ as follows:
  \begin{enumerate}
    \item If of~$uv,vu \in \dir G_{2r}$,
          then orient~$uv$ in~$\dir H$ according to the corresponding
          arc in $\dir G_{2r}$ (if both arcs exists choose an arbitrary
          orientation),
    \item otherwise there exists~$w \in N^-_{2r}(u) \cap N^-_{2r}(v)$
          with~$\omega_{2r}(u) + \omega_{2r}(v) = \dist_G(u,v) \leq 2r$. Orient the edge~$uv$ towards that vertex~$x \in \{u,v\}$
          for which~$\omega_{2r}(x)$ is larger.
  \end{enumerate}
  We now argue that~$\maxindeg(\dir H)$ is small. Consider any vertex $v \in
  \dir H$. Every in-arc $uv \in \dir H$ either is of type~1, of which we have
  at most~$\maxindeg(\dir G_{2r})$, or of type~2. Consider a group of in-
  arcs~$u_iv$, $1 \leq i \leq \ell$ of type~2 that are all present because of
  a common vertex~$w$. Since~$w \in N^-_{2r}(u)$, we have at
  most~$\maxindeg(\dir G_{2r})$ such groups. By construction,
  $\omega_{2r}(wu_i) \leq \omega_{2r}(wv)$ and since both weights sum to less
  than~$2r$, this means that~$\omega_{2r}(wu_i) \leq r$.
  Lemma~\ref{lemma:D1-intersection} now tells us that~$\ell \leq \tau + 1$.
  Therefore~$v$ has at most~$(\tau+1) \maxindeg(\dir G_{2r})$ in-arcs of type~2 and we conclude that
  \[
    \maxindeg(\dir H) \leq \maxindeg(\dir G_{2r}) + (\tau+1)\maxindeg(\dir G_{2r}) = (\tau+2)\maxindeg(\dir G_{2r}).
  \]
  This finally implies that~$H$ is $2(\tau+2)\maxindeg(\dir G_{2r})$-degenerate and therefore contains an independent set~$A \subseteq V(H)$
  of size at least~$|A| \geq |H| / (2(\tau+2)\maxindeg(\dir G_{2r}))$.
  Taken together with the fact that~$|H| = |D_1| \geq |D| / \maxindeg(\dir G_r)$
  (every vertex added to~$D_1$ will cause at most~$\maxindeg(\dir G_r)$ many
  vertices to be added to~$D_2$ in the loop at line~\ref{algstep:pushloop}
  and~$D = D_1 \cup D_2$), we find that
  \[
  	|A| \geq \frac{|D|}{2(\tau+2)\maxindeg(\dir G_{2r}))\maxindeg(\dir G_r)}
  \]
  By construction of~$H$ we conclude that~$A$ is $(2r+1)$-scattered
  in~$G$ of the claimed size.
\end{proof}

\noindent
Since a $(2r+1)$-scattered set provides a lower bound for an~$r$-dominating
set, we conclude that Algorithm~\ref{alg:r-domset} computes a
$2(\tau+2)\maxindeg(\dir G_{2r})\maxindeg(\dir G_r)$-approximation of an
optimal $r$-dominating set. In other words, we obtain a constant-factor
approximation in graphs of bounded expansion since the quantities $\maxindeg(\dir
G_{2r})\maxindeg(\dir G_r)$ are constants per Theorem~\ref{thm:dtf} in bounded
expansion classes and the quantity $2(\tau+2)$ is a constant of our choosing.

In practice one could, depending on the value of $\maxindeg(\dir G_r)$
and $\maxindeg(\dir G_{2r})$, compute the optimal value for~$\tau$ to minimize the 
approximation guarantee. However, this
would necessitate the computation of $2r$ augmentation, the expensive step we
want to avoid. Alternatively, we can choose a `good enough' value for $\tau$
that still guarantees a constant-factor approximation while being easy to
determine in practice. In the context of cDBGs, we found that
$\tau := (2r)^2$ yields reliably good results.

\subsection{Computational Runtimes}
\label{subsec:runtimes}

See ``Benchmarking'' in Materials and Methods for benchmarking methods.

The \podarv data set was retrieved from the NCBI SRA using accession
SRR606249.  The full build and indexing of the 103 million
error-trimmed reads (10.3 Gbp in total) took approximately 23 minutes
and required 12.8 GB of RAM. Loading the indices for search required
4.3 GB of RAM and a search with a 3 Mbp genome took approximately 32
seconds.

The \hu data set was retrieved from the NCBI SRA using accession
SRR1976948. The full build and indexing of the 34 million
error-trimmed reads (8.5 Gbp in total) required approximately 217
minutes and required 24.4 GB of RAM. Loading the indices for search
required 18 GB of RAM and a search with a 3 Mbp genome took
approximately 80 seconds.

For data set complexity (number of k-mers, number of cDBG nodes) please see
Table~\ref{tab:data}.

\subsection{spacegraphcats pipeline overview}
\label{subsec:sgc_pipeline}

\textsf{spacegraphcats} follows a series of steps when run on sequencing
data, see Figure~\ref{fig:sgc_dag}. In detail, we perform the following
steps.

\emph{BCALM}. Use BCALM to generate a cDBG. Then convert a BCALM {\tt unitigs.fa}
output (a cDBG) into spacegraphcats files.
Outputs an undirected graph, a file containing the sequences, and a
{\tt .info.csv} file containing information about the contig. Also outputs 
sourmash {\tt k=31}, {\tt scaled=1000} signatures for both input and
output files.

\emph{spacegraphcats.cdbg.label\_cdbg}. Build an index that can be
used to retrieve individual reads or contigs by cDBG node ID; produce 
a SQLite database for fast retrieval. Briefly, this script creates a 
sqlite database with a single table, {\tt sequences}, where a query like
{\tt SELECT DISTINCT sequences.offset FROM sequences WHERE label ...}
can be executed to return the offset of all sequences with the given
label; the offsets refer to BGZF coordinates in the gzipped sequence collection. Here, 'label' is the cDBG ID to which the sequence belongs.

The script {\tt extract\_reads\_by\_frontier\_sqlite.py} is a downstream script to
extract the reads with a frontier search.
Specifically: 1. walk through the contigs assembled from the cDBG;  
2. build a DBG cover using khmer tags, such that every k-mer in the DBG is 
within distance d=40 of a tag;  3. label each tag with the cDBG node ID 
from the contig; 4. save for later use.

\emph{spacegraphcats.catlas.catlas}. The catlas is a hierarchical 
atlas for querying graphs. Implements algorithms 1 and 2 (see main text).

\emph{spacegraphcats.index.index\_contigs\_by\_kmer}. Use Minimal Perfect 
Hashing (BBHash, https://github.com/rizkg/BBHash) to construct a fast
lookup table connecting k-mers in the cDBG to cDBG node IDs. (BBHash
reference: A. Limasset, G. Rizk, R. Chikhi, P. Peterlongo, Fast and
Scalable Minimal Perfect Hashing for Massive Key Sets, SEA 2017.)

\emph{spacegraphcats.search.extract\_nodes\_by\_query}. Do a frontier 
search, and retrieve cDBG node IDs and MinHash signature for the retrieved 
contigs.

\emph{spacegraphcats.search.extract\_contigs}. Retrieve the unitig sequences 
for a given list of cDBG nodes.  Consumes the output of {\tt extract\_nodes\_by\_query} 
to get the list of nodes.

\emph{spacegraphcats.search.extract\_reads}. Retrieve the reads for a list of cDBG 
nodes.  Consumes the output of {\tt extract\_nodes\_by\_query} to get the list of 
nodes, and then uses the labeled cDBG output by {\tt .cdbg.label\_cdbg} to 
find reads that overlap with the unitigs in those nodes.
  

\begin{figure*}
 \centering
 \includegraphics[width=\linewidth]{figures/sgc_dag}
	\caption{The steps followed by \textsf{spacegraphcats} when run
on sequencing data.
}
 \label{fig:sgc_dag}
\end{figure*}

\subsection{Query genome accession numbers for {\em Proteiniclasticum} search}
\label{subsec:query_accessions}

See Table~\ref{tab:query_accessions}.

\begin{table}[b]
  \begin{tabular}{l l c c }
    \toprule
    Name & NCBI accession \\
    \midrule
    \hline
    P. ruminis CGMCC & GCA\_900099635.1 \\
    P. ruminis DSM & GCA\_000701905.1 \\
    P. ruminis ML2 & GCA\_900115135.1 \\
    \hline
    \bottomrule
  \end{tabular}
  \caption{Accession numbers for genomes used in {\em Proteiniclasticum} neighborhood query.}
  \label{tab:query_accessions}
\end{table}

\subsection{Amino Acid Identity results for {\em Proteiniclasticum}}
\label{subsec:aai}

See Table~\ref{tab:aai}.

\begin{table}
  \begin{tabular}{l l c c }
    \toprule
    Genome A & Genome B & Orthologous Genes & Mean AAI \\
    \midrule
    {\em P. ruminis ML2} & {\em P. ruminis shakya} & 2546 & 95.74 \\
    {\em P. ruminis DSM} & {\em P. ruminis shakya}  & 2391 & 93.47 \\
    \hline
    \bottomrule
  \end{tabular}
  \caption{CompareM results for {\em Proteiniclasticum} genomes. {\em P. ruminis shakya} is the result of assembling the reads extracted from \podarv with the neighborhood search.}
  \label{tab:aai}
\end{table}


\subsection{\hu analysis pipeline overview}


See Figure~\ref{fig:hu_dag}. We implemented three workflows to analyze the
plass-assembled \hu query neighborhoods. 


\begin{figure*}
 \centering
 \includegraphics[width=\linewidth]{figures/hu_dag}
	\caption{Three workflows implemented to analyze the plass-assembled
\hu query neighborhoods. The first three steps, depicted in blue, were common
across all workflows. The green boxes depict the KEGG GhostKOALA annotation
workflow, the results of which can be see in Figure~\ref{fig:hu_content}. 
The orange boxes depict steps in common between the clustering and variant
workflows used to generate Figure~\ref{fig:hu_content}. The red boxes depict
steps used to generate the MDS clustering plot and the multiple sequence 
alignment (see Figure~\ref{fig:gyrAalign}). The gold boxes depict the
steps of the variant workflow used to generate the assembly graphs.  
}
 \label{fig:hu_dag}
\end{figure*}

\subsection{Genome bin completeness improvements for \hu}
\label{subsec:checkm}

See Table~\ref{tab:completeness}.

\begin{table*}
  \parbox[t][][t]{.55\linewidth}{%
  \begin{tabular}{@{}l c c c@{}}
    \toprule
    Bin name & Bin completeness & MEGAHIT ($\Delta$) & \plass ($\Delta$) \\
    \midrule
    % generated from notebook 'figures/fig-hu_strain.ipynb'
    \input{tables/bin-completeness.tex}
 \\
    \bottomrule
  \end{tabular}
  \caption{\label{tab:completeness}%
  Bin and neighborhood completeness, as estimated by CheckM. ``Bin completeness''
  is the result of running CheckM on the genome sequence from GenBank; MEGAHIT
   is the result of running CheckM on the MEGAHIT nucleotide assembly of the
   neighborhood reads; \plass is the result of running CheckM on the \plass
   amino acid assembly of the neighborhood reads. $\Delta$ is the difference
   between the column and the bin completeness.}
  }%
  \hfill
  \parbox[t][][t]{.4\linewidth}{
  \begin{tabular}{@{}l l c c @{}}
    \toprule
    Species & gyrA (bin) & gyrA (\plass) \\
    \midrule
    % generated from notebook 'tables/plass-genes/distinct_positions.ipynb'
    Methanobacterium sp. 42\_16 & 0 & 0 \\
    P. bacterium 34\_609 & 0 & 0 \\
    Desulfotomaculum sp. 46\_80 & 0 & 0 \\
    S. bacterium 57\_84 & 0 & 0 \\
    B. bacterium & 0 & 1 \\
    P. acetatigenes isolate 50\_10 & 0 & 2 \\
    WS6 bacterium 34\_10 & 0 & 2 \\
    M. marisnigri isolate 62\_101 & 0 & 2 \\
    C. bacterium 38\_11 & 1 & 1 \\
    M. infera isolate 46\_47 & 1 & 1 \\
    S. bacterium 53\_16 & 1 & 1 \\
    M. bacterium 46\_47 & 1 & 1 \\
    TA06 bacterium 32\_111 & 1 & 1 \\
    P. bacterium 33\_209 & 1 & 1 \\
    A. bacterium 66\_15 & 1 & 1 \\
    Methanocalculus sp. 52\_23 & 1 & 2 \\
    WS6 bacterium 36\_33 & 1 & 2 \\
    A. bacterium 34\_128 & 1 & 2 \\
    A. thermophila isolate 46\_16 & 1 & 2 \\
    M. harundinacea isolate 57\_489 & 1 & 2 \\
    M. bacterium 39\_7 & 2 & 0 \\
    Desulfotomaculum sp. 46\_296 & 2 & 2 \\
    A. bacterium 49\_20 & 2 & 3 \\
    \bottomrule
  \end{tabular}
  \caption{  \label{tab:gyrAcounts}%
   Bin and neighborhood gyrA protein content. gyrA count for each bin is the number
    of gyrA amino acid sequences that are part of the original bin. gyrA count by \plass is
    the minimum number of gyrA amino acid sequences supported by at least one
    position with at least 10 copies of each variant, e.g. ``3'' indicates that
    there is at least one position in the multiple sequence alignment of gyrA
    sequences for that neighborhood that has 3 distinct variants in 10 distinct
    sequences.}
  }
\end{table*}

\subsection{K-mer inclusion of reads by MEGAHIT assemblies}
\label{subsec:inclusion}

See Table~\ref{tab:kmer_inclusion}. We estimated the number of k-mers in each
query neighborhood that were contained in the MEGAHIT assembly of that query
neighborhood. We used sourmash {\tt compute} to calculate signatures with k-size of
31 and a scaled value of 2000. We then used sourmash {\tt compare} to estimate
containment in MEGAHIT assemblies. The query neighborhood with the smallest
containment, \emph{M. harundinacea} isolate \emph{57\_489}, had the largest query
neighborhood, while the query neighborhood with the largest containment,
\emph{M. bacterium 39\_7}, had the smallest query neighborhood.

\begin{table}
  \begin{tabular}{l c}
    \toprule
    Species & MEGAHIT assembly containment \\
    \midrule
    % generated from notebook 'figures/megahit-assembly-inclusion.ipynb'
    \input{tables/megahit-inclusion.table.tex}
%    \input{tables/plass-genes/rpsC-table.tex}
    \bottomrule
  \end{tabular}
  \caption{Containment of neighborhood k-mer content in MEGAHIT nucleotide assemblies.}
  \label{tab:kmer_inclusion}
\end{table}

\subsection{gyrA alignment}
\label{subsec:gyrAalign}

See Figure~\ref{fig:gyrAalign}. The MDS plot in the left panel of figure 4 shows
distinct gyrA sequences identified in the Plass assemblies using HMMER. To visualize
the sequences within these clusters and in other query neighborhoods, we
constructed a multiple sequence alignment. However, because many
sequences assembled by Plass were fragmented (see Results: Some query neighborhoods
contain substantial strain variation), we first clustered the sequences at 95\%
similarity using CD-HIT. We then aligned the centroid sequences using MAFFT with
default settings. To produce the multiple sequence alignment visualization, we
calculated an unrooted neighbor joining tree using the MAFFT alignment. Then we used
the function {\tt msaplot} in the R package ggtree to plot the alignment.

\begin{figure*}
 \centering
 \includegraphics[width=\linewidth]{figures/gyrA-cdhit95-msa}
	\caption{A multiple sequence alignment and neighbor joining tree of representative gyrA amino acid fragments assembled by \plass from the genome neighborhoods in \hu. Protein sequences that originated from the genome bin are prepended with "Bin." All other sequences were assembled by Plass.
 }
 \label{fig:gyrAalign}
\end{figure*}

\subsection{gyrA by neighborhood}
\label{subsec:gyrAnbhd}

See Table~\ref{tab:gyrAcounts}. As can be seen in the left panel of figure 4 in the
main text, we observe many unique amino acid sequences per single copy ortholog
per query neighborhood.
Although we observe many possible traversal paths in compact De Bruijn graphs built
from reads that give rise to these sequences, we have no way to ascertain whether we
observed combinatorial complexity by assembling variants that would never be linked
in nature. Therefore, we sought to conservatively estimate the number of positions
per amino acid sequence that contained variants using MAFFT alignments. First, we
subsetted the alignment to sequences from one query neighborhood. Then we identified
all aligned non-gap characters for each position in the alignment (gaps were induced
in some neighborhoods by the presence of amino acid residues in other query neighborhood
amino acid sequences). For each of these positions, we counted the number of unique
amino acid sequences per position, and the number of times each occurred at that position.
We then elimated any variant that occurred fewer than 10 times. Lastly, we counted the
number of well-supported distinct characters. We did this for gyrA, as well as the
amino acid sequences for the other genes we tested (see other genes). Table~\ref{tab:gyrAcounts}
shows that we see increased number of gyrA sequences in many neighborhoods even with
this conservative approach.

\subsection{Other genes}

\label{subsec:othergenes}

See bin and neighborhood content results for alaS in Table~\ref{tab:alaS}, {\em gyrB} in
Table~\ref{tab:gyrB}, {\em recA} in Table~\ref{tab:recA}, {\em rpb2d6} in Table~\ref{tab:rpb2d6},
rplB in Table~\ref{tab:rplB}, and rpsC in Table~\ref{tab:rpsC}. We selected {\em gyrB} and
{\em recA} because they were used by \hu to assign taxonomy to binned genomes. We selected
other genes used as single copy orthologs by programs like CheckM, and with longer PFAM
domains.


\begin{table}
  \begin{tabular}{l l c c }
    \toprule
    Name & PFAM accession \\
    \midrule
    recA & PF00154 \\
    rplB & PF00181 \\
    rpsC & PF00189 \\
    gyrB & PF00204 \\
    gyrA & PF00521 \\
    rpb2d6 & PF00562 \\
    alaS & PF01411 \\
    \hline
    \bottomrule
  \end{tabular}
  \caption{Protein names and PFAM accessions for targeted analyses.}
  \label{tab:pfam_accessions}
\end{table}

\begin{table}
  \begin{tabular}{l l c c }
    \toprule
    Species & alaS (bin) & alaS (\plass) \\
    \midrule
    % generated from notebook 'tables/plass-genes/distinct_positions.ipynb'
    \input{tables/plass-genes/alaS-table.tex}
    \bottomrule
  \end{tabular}
  \caption{Bin and neighborhood alaS protein content.}
  \label{tab:alaS}
\end{table}

\begin{table}
  \begin{tabular}{l l c c }
    \toprule
    Species & gyrB (bin) & gyrB (\plass) \\
    \midrule
    % generated from notebook 'tables/plass-genes/distinct_positions.ipynb'
    \input{tables/plass-genes/gyrB-table.tex}
    \bottomrule
  \end{tabular}
  \caption{Bin and neighborhood gyrB protein content.}
  \label{tab:gyrB}
\end{table}

\begin{table}
  \begin{tabular}{l l c c }
    \toprule
    Species & recA (bin) & recA (\plass) \\
    \midrule
    % generated from notebook 'tables/plass-genes/distinct_positions.ipynb'
    \input{tables/plass-genes/recA-table.tex}
    \bottomrule
  \end{tabular}
  \caption{Bin and neighborhood recA protein content.}
  \label{tab:recA}
\end{table}

\begin{table}
  \begin{tabular}{l l c c }
    \toprule
    Species & rpb2d6 (bin) & rpb2d6 (\plass) \\
    \midrule
    % generated from notebook 'tables/plass-genes/distinct_positions.ipynb'
    \input{tables/plass-genes/rpb-table.tex}
    \bottomrule
  \end{tabular}
  \caption{Bin and neighborhood rpb2d6 protein content.}
  \label{tab:rpb2d6}
\end{table}

\begin{table}
  \begin{tabular}{l l c c }
    \toprule
    Species & rplB (bin) & rplB (\plass) \\
    \midrule
    % generated from notebook 'tables/plass-genes/distinct_positions.ipynb'
    \input{tables/plass-genes/rplB-table.tex}
    \bottomrule
  \end{tabular}
  \caption{Bin and neighborhood rplB protein content.}
  \label{tab:rplB}
\end{table}

\begin{table}
  \begin{tabular}{l l c c }
    \toprule
    Species & rpsC (bin) & rpsC (\plass) \\
    \midrule
    % generated from notebook 'tables/plass-genes/distinct_positions.ipynb'
    \input{tables/plass-genes/rpsC-table.tex}
    \bottomrule
  \end{tabular}
  \caption{Bin and neighborhood rpsC protein content.}
  \label{tab:rpsC}
\end{table}
